static/src/javascripts/bootstraps/enhanced/notifications.js

```diff
@@ -0,0 +1,180 @@
+    template,
+    mediator,
+    robust,
+    svgs,
+    userPrefs,
+    subscribeTemplate,
+    some,
+    uniq,
+    without
+) {
+    var modules, reg, sub,
+        isSubscribed = false;
+
+    modules = {
+
+        getPushSubscription: function () {
+
+            navigator.serviceWorker.ready.then(function (serviceWorkerRegistration) {
+                reg = serviceWorkerRegistration;
+                serviceWorkerRegistration.pushManager.getSubscription().then(function (pushSubscription) {
+
+                    if (pushSubscription) {
+                        sub = pushSubscription;
+                        modules.configureSubscribeTemplate();
+                    } else {
+                        reg.pushManager.subscribe({userVisibleOnly: true}).then(
+                            function (pushSubscription) {
+                                sub = pushSubscription;
+                                modules.configureSubscribeTemplate();
+                            }
+                        );
+                    }
+                });
+            });
+        },
+
+        configureSubscribeTemplate: function () {
+            var subscribed = modules.checkSubscriptions(),
+                hasNoSubscriptions = modules.subscriptionsEmpty(),
+                src = template(subscribeTemplate, {
+                    className: hasNoSubscriptions ? '' : 'notifications-subscribe-link--has-subscriptions',
+                    text: subscribed ? 'Following story' : 'Follow story',
+                    imgMobile: svgs('notificationsExplainerMobile', ['mobile-only', 'notification-explainer']),
+                    imgDesktop: svgs('notificationsExplainerDesktop', ['hide-on-mobile', 'notification-explainer'])
+                });
+
+            fastdom.write(function () {
+                $('.js-liveblog-body').prepend(src);
+                mediator.emit('page:notifications:ready');
+            });
+
+            bean.on(document.body, 'click', '.js-notifications-subscribe-link', function () {
+                modules.buttonHandler();
+            });
+        },
+
+        setSubscriptionStatus: function (subscribed) {
+            var subscribeButton = $('.js-notifications-subscribe-link'),
+                subscribeLink = $('.notifications-subscribe-link--follow');
+            isSubscribed = subscribed;
+            if (subscribed) {
+                subscribeLink.addClass('notifications-subscribe-link--has-subscriptions');
+            }
+            subscribeButton[0].textContent = subscribed ?  'Following story' : 'Follow story';
+        },
+
+        buttonHandler: function () {
+            if (isSubscribed) {
+                modules.stopFollowingThis();
+            } else {
+                modules.subscribe();
+            }
+        },
+
+        subscribe: function () {
+            if (modules.subscriptionsEmpty()) {
+                reg.pushManager.subscribe({userVisibleOnly: true}).then(function (pushSubscription) {
```

Comment:
I think it can be quite confusing when you depend on outside variables (not constants) in this way. I will try to explain why!

Looking at this code for the first time, it's not clear where `reg` comes from. Upon inspection, you realise it's a variable which is initially set to `undefined`. This function assumes that the `getPushSubscription` will run prior to this, as that function, if successful, will reset the value of the `reg` variable to the type `ServiceWorkerRegistration`.

Would this function be easier to read and safer if it didn't have that as a dependency? This way it is explicit what the function depends on, and there's no surprises, whereas before you never knew what `reg` was (was it undefined or an object?). Perhaps something like:

``` js
getReg: function () {
    return navigator.serviceWorker.ready;
}),
subscribe: function () {
    if (modules.subscriptionsEmpty()) {
        getReg()
            .then(function (reg) { return reg.pushManager.subscribe({userVisibleOnly: true}) })
            .then(function (pushSubscription) {
```

Likewise, we are doing the same thing with `sub`. Instead of having a global variable which changes over time, would it be more readable if functions which depended on `sub` explicitly asked for it?

You could then replace all instances of `reg` and `sub` with usages of these functions:

``` js
getReg: function () {
    return navigator.serviceWorker.ready;
}),
getSubscription: function () {
    return getReg().then(function (reg) { return reg.pushManager.getSubscription() });
}
```

Hope this is clear! Happy to discuss in person if it's easier :-) Let me know what you think!